src/lib/protocols/collectd.c

```diff
@@ -0,0 +1,209 @@
+                                              u_int16_t block_offset, u_int16_t block_length)
+{
+  return (ndpi_hostname_sni_set(flow, &packet->payload[4], block_length) == NULL);
+}
+
+static int ndpi_int_collectd_dissect_username(struct ndpi_flow_struct * const flow,
+                                              struct ndpi_packet_struct const * const packet,
+                                              u_int16_t block_offset)
+{
+  u_int16_t username_length = ntohs(get_u_int16_t(packet->payload, 4));
+
+  if(username_length > packet->payload_packet_len -
+     COLLECTD_ENCR_AES256_MIN_BLOCK_SIZE -
+     COLLECTD_ENCR_AES256_IV_SIZE)
+  {
+    return 1;
+  }
+
+  size_t sz_len = ndpi_min(sizeof(flow->protos.collectd.client_username) - 1, username_length);
+  memcpy(flow->protos.collectd.client_username, &packet->payload[6], sz_len);
+  flow->protos.collectd.client_username[sz_len] = '\0';
+
+  return 0;
+}
+
+void ndpi_search_collectd(struct ndpi_detection_module_struct *ndpi_struct,
+                          struct ndpi_flow_struct *flow)
+{
+  struct ndpi_packet_struct const * const packet = &ndpi_struct->packet;
+  u_int16_t num_blocks;
+  u_int16_t block_offset = 0, block_type, block_length;
+  u_int16_t hostname_offset, hostname_length = 0;
+
+  NDPI_LOG_DBG(ndpi_struct, "search collectd\n");
+
+  for (num_blocks = 0; num_blocks < COLLECTD_MAX_BLOCKS_TO_DISSECT;
+       ++num_blocks, block_offset += block_length)
+  {
+    block_length = npdi_int_collectd_block_size(packet, block_offset);
+    if (block_length == 0)
+    {
+      break;
+    }
+
+    block_type = ntohs(get_u_int16_t(packet->payload, block_offset));
+    if (ndpi_int_collectd_check_type(block_type) != 0)
+    {
+      break;
+    } else {
+      if (block_type == COLLECTD_TYPE_HOST)
+      {
+        /*
+         * Dissect the hostname later, when we are sure that it is
+         * the collectd protocol.
+         */
+        hostname_offset = block_offset;
+        hostname_length = block_length;
+      } else if (block_type == COLELCTD_TYPE_ENCR_AES256) {
+        /*
+         * The encrypted data block is a special case.
+         * It is the only dissectable block as everything else in it
+         * is encrypted.
+         */
+        if (block_length != packet->payload_packet_len ||
+            block_length < COLLECTD_ENCR_AES256_MIN_BLOCK_SIZE ||
+            ndpi_int_collectd_dissect_username(flow, packet, block_offset) != 0)
+        {
+          NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
+        } else {
+          ndpi_int_collectd_add_connection(ndpi_struct, flow);
+        }
+        return;
+      }
+    }
+  }
+
+  if (num_blocks < COLLECTD_MIN_BLOCKS_REQUIRED)
+  {
+    NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
+    return;
+  }
+
+  if (hostname_length > 0 &&
+      ndpi_int_collectd_dissect_hostname(flow, packet, hostname_offset,
+                                         hostname_length) != 0)
```

Comment:
Just a nit: `ndpi_int_collectd_dissect_hostname` never fails, because `ndpi_hostname_sni_set` never return NULL; so this check is useless.
Up to you if removing the check or not; I am fine in both cases

Reply:
I know, but I think it makes sense to extend `ndpi_hostname_sni_set` at some point e.g. so that it will set a risk if a non-printable character is seen.